FEA Sequential feature selection

rasbt opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->

#6545

#### What does this implement/fix? Explain your changes.

Implements a new class for sequential feature selection

#### Any other comments?

This is a first PR to pick-up the discussion from #6545 and to get some feedback for improvements. 
Also, there are some more todos I have in mind when we get further down this PR:

Todos:
- [ ] Add documentation & examples
- [ ] Add more unit tests (e.g., GridSearchCV)


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

rasbt committed:
sequential feature selection

rasbt committed:
sequential feature selection

rasbt commented:
the unit tests are failing partly because of the non-meta estimator checks, e.g.,

```
======================================================================
ERROR: /home/travis/build/scikit-learn/scikit-learn/sklearn/tests/test_common.py.test_non_meta_estimators:check_fit1d_1sample(SFS)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/utils/testing.py", line 741, in __call__
    return self.check(*args, **kwargs)
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/utils/testing.py", line 292, in wrapper
    return fn(*args, **kwargs)
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/utils/estimator_checks.py", line 640, in check_fit1d_1sample
    estimator = Estimator()
TypeError: __init__() missing 1 required positional argument: 'estimator'```

What would be a good way to handle this particular case, since the SFS would require a positional argument `estimator` that is used to perform the feature selection with. Do you think it's a good idea to use a default for that (e.g., logistic regression for classification and linear regression or regression), or would this SFS be better suited as a meta-estimator?

chkoar commented on sklearn/feature_selection/sfs.py:
```diff
@@ -0,0 +1,366 @@
+        self.forward = forward
+        self.pre_dispatch = pre_dispatch
+        self.scoring = scoring
+        if scoring is None:
```
The `scikit-learn`'s convention is not to alter parameters in `__init__` method. Validation must be performed in `fit`. The `Pipeline` object is an exception I think.

chkoar commented on sklearn/feature_selection/sfs.py:
```diff
@@ -0,0 +1,366 @@
+        self.pre_dispatch = pre_dispatch
+        self.scoring = scoring
+        if scoring is None:
+            if self.estimator._estimator_type == 'classifier':
```
You could add this to get the type
```python
    @property
    def _estimator_type(self):
return self.estimator._estimator_type
```

GaelVaroquaux commented:
You need to avoid using "SFS", as it's hard to know what it means. Maybe
"SeqFeatureSelection"?

chkoar commented on sklearn/feature_selection/sfs.py:
```diff
@@ -0,0 +1,366 @@
+                raise ValueError('Estimator must '
+                                 'be a Classifier or Regressor.')
+
+        if isinstance(scoring, str):
```
You could use the `check_scoring` from `sklearn.metrics.scorer ` in the `fit` method to get the `scorer`

chkoar commented on sklearn/feature_selection/sfs.py:
```diff
@@ -0,0 +1,366 @@
+    ----------
+    estimator : scikit-learn Classifier or Regressor
+
+    k_features : int or tuple (default=1)
```
I would name it `n_features_to_select` like in RFE.

chkoar commented on sklearn/feature_selection/sfs.py:
```diff
@@ -0,0 +1,366 @@
+from ..metrics import get_scorer
+
+
+class SFS(BaseEstimator, MetaEstimatorMixin):
```
I would expect to derive from `SelectorMixin` like RFE

rasbt replied:
agreed! Tried that first though and was having some issues, i.e. `TypeError: Can't instantiate abstract class SFS with abstract methods _get_support_mask`. Do you know what would need to be added to make this work?

chkoar replied:
You should implement the `_get_support_mask` method. This method must return a boolean array of shape [# input features]. (True if its corresponding feature is selected False otherwise). So, you will get the `transform` method from  `SelectorMixin`.

chkoar commented:
Hey @rasbt,  thanks for this contribution. I would name it `SequentialFeatureSelector` or `SequentialFeatureSelection`. It is not that long and clearly states what you would expect from that class.

TomDLT commented:
> would this SFS be better suited as a meta-estimator?

I would say so. You can add it here:
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/testing.py#L468

rasbt commented:
Thanks for the feedback everyone. Will try to address all of your points soon and get an updated PR up. 

@Gael @chkoar 
> You need to avoid using "SFS", as it's hard to know what it means. Maybe
"SeqFeatureSelection"?
...
> would name it SequentialFeatureSelector or SequentialFeatureSelection

Completely agree with you both. I would prefer `SequentialFeatureSelector` (that's also how I called the [orig](http://rasbt.github.io/mlxtend/user_guide/feature_selection/SequentialFeatureSelector/) implementation) -- if you think that's not too long, @GaelVaroquaux . Otherwise, we could use `SeqFeatureSelection` / `SeqFeatureSelector` like you suggested. (I went with SFS because of RFE, but I agree, no one would now what SFS is :))

rasbt commented:
> > would this SFS be better suited as a meta-estimator?

> I would say so. You can add it here:

Thanks @TomDLT , will make it a meta-estimator :) --  makes sense since it is a "wrapper algorithm" for feature selection.

rasbt committed:
addressed comments by chkoar

rasbt committed:
addressed comments by chkoar

rasbt committed:
fix doctests and mask

rasbt committed:
fix doctests and mask

rasbt committed:
fix estimator cloning

rasbt committed:
fix estimator cloning

rasbt committed:
change k_features to feature_subset

rasbt committed:
change k_features to feature_subset

rasbt committed:
make compatible to old numpy versions in ci

rasbt committed:
make compatible to old numpy versions in ci

rasbt committed:
flake8 checks

rasbt committed:
flake8 checks

rasbt committed:
flake8 checks

rasbt committed:
flake8 checks

GaelVaroquaux commented:
> Completely agree with you both. I would prefer SequentialFeatureSelector
> (that's also how I called the orig implementation) -- if you think that's not
> too long, @GaelVaroquaux .

I am OK with that.

> I went with SFS because of RFE,

I am not sure that today we would name RFE like that, with insight.

rasbt commented:
Sounds good! 

In the meantime, I addressed all your code suggestions, @chkoar -- thanks for the very useful feedback! The unit tests seem to pass now as well. Let me know if you have further suggestions. I will try to get to the documentation in the next couple of days.

glemaitre commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,323 @@
+        self.scoring = scoring
+        self.cv = cv
+        self.n_jobs = n_jobs
+        self.subsets_ = {}
```
This should go in `fit`, isn't it?

chkoar replied:
Conventionally, yes.

rasbt replied:
good catch! will fix that in a sec

rasbt committed:
remove duplicate self.subsets_

rasbt committed:
remove duplicate self.subsets_

chkoar commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,322 @@
+        self : object
+
+        """
+        if self.scoring is None:
```
I think that we could remove these four lines, no?
```python
check_scoring(self.estimator. scoring=self.scoring)
```

chkoar replied:
Sorry for my ignorance, but is there a specific reason for attaching the scorer in the instance itself?

rasbt replied:
I think if we would remove those lines, the evaluation metric would always be either 'accuracy' (for classifiers) or 'r2' (for regressors). I think these are good as default, but user may prefer different metrics for the feature subset selection.

chkoar replied:
By using this one 
```python
check_scoring(self.estimator. scoring=self.scoring)
```
a user could pass whatever metric he/she wants

rasbt replied:
Oh nice, didn't know that! In this case, we can remove those lines, of course.

> specific reason for attaching the scorer in the instance itself?

It's used in several helper methods, so it would be easier to fetch the once instead of passing it around as function argument to the helper methods I guess. I renamed it to `self._scorer` but we can also completely remove it, just thought that it makes the code a bit more verbose

rasbt committed:
simplify scorer

rasbt committed:
simplify scorer

jorisvandenbossche commented:
I would like to see this added to scikit-learn. @rasbt Is there anything I can help with, or does it mainly need review of core developers?

glemaitre commented:
It looks something that could be pushed during the sprint (I am starting to make a list ;))

rasbt commented:
I am happy to make any modifications if necessary. Other than that, I think it's just the documentation that's missing. I would take the one at http://rasbt.github.io/mlxtend/user_guide/feature_selection/SequentialFeatureSelector/ as a template and simplify/modify it a bit so that it is consistent with the code as it is currently implemented. 

@jorisvandenbossche I don't have anything particular in mind regarding potential changes, but if you have any comments regarding to code, please let me know.

austinmw commented:
Patiently waiting for forward selection based on R2 and/or p-vals to be added 😊

rasbt commented:
Unless there's a bug, it should already support 'R2'-based feature selection. Regarding the p-values, I don't think adding this is a good idea because approaches like sequential feature selection and k-fold selection would violate assumptions in typical statistical hypothesis tests I'd say

rasbt committed:
support for scorers that return nan; bugfix and additional unittest to select features in a specified range

rasbt committed:
support for scorers that return nan; bugfix and additional unittest to select features in a specified range

amueller commented:
mrg conflict? CI error? ;)

rasbt commented:
> mrg conflict? CI error? ;)

ehm, yeah ... everything under control now, has been such a long time since I started this, and it desperately needed a rebase

rasbt committed:
pep8 style fix

rasbt committed:
pep8 style fix

rasbt commented:
I was about to set up the documentation for this `SFS` class and remembered that there are currently two implementations for recursive feature elimination: `RFE` and `RFECV`. Now, I was wondering if this `SFS` class should be split up in the same way, i.e., having an `SFS` and and `SFSCV`? Or would that be overkill?

(Note that the current `SFS` implementation has a `cv` argument like `RFECV`, this class would hence become `SFSCV`). 

@glemaitre @chkoar @amueller , what do you think?

jnothman commented:
I think a single class would be acceptable. I think, but I might be wrong,
that RFECV is a bit different in that it performs a full RFE fit for each
train-test split then aggregates over their results, rather than making
each evaluation using CV.

On 18 July 2017 at 10:17, Sebastian Raschka <notifications@github.com>
wrote:

> I was about to set up the documentation for this SFS class and remembered
> that there are currently two implementations for recursive feature
> elimination: RFE and RFECV. Now, I was wondering if this SFS class should
> be split up in the same way, i.e., having an SFS and and SFSCV? Or would
> that be overkill?
>
> (Note that the current SFS implementation has a cv argument like RFECV,
> this class would hence become SFSCV).
>
> @glemaitre <https://github.com/glemaitre> @chkoar
> <https://github.com/chkoar> @amueller <https://github.com/amueller> ,
> what do you think?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8684#issuecomment-315920512>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz66-uQuWxX1vAVpOW1879w0phXze2ks5sO_gUgaJpZM4MwYuz>
> .
>

rasbt commented:
Yeah you are right, SFS & RFE are really doing different things. Since SFS evaluates subsets based on estimator performance, I think it wouldn't make much sense to have an SFS without `cv` -- it would hopelessly overfit.

rasbt commented:
regarding the last commit; used the "resolve merge conflict" feat in GitHub UI, which was probably not ideal (hence, the weird name of that commit "Merge branch 'master' into sfs"). In any case, just updated the What's new" doc to the recent one in master to avoid merge conflicts. Will modify it later, adding the SFS to the document once the changes are all in place (currently, I think there are too many changes going on since 0.19 is coming out)

Haybu commented:
what is the status of this PR? is it released?

rasbt commented:
I think it's still in a "discussion" or "planning" stage regarding the general API layout (esp. wrt. CV). For now, you can use the `SequentialFeatureSelector` from mlxtend until this PR gets figures out :) (http://rasbt.github.io/mlxtend/user_guide/feature_selection/SequentialFeatureSelector/)

jnothman commented:
it's in my pile! (but so is a lot else)

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,323 @@
+        (0, 2, 3)
+        >>> sfs.transform(X).shape
+        (150, 3)
+
```
See also RFE, rfecv, SelectFromModel and reciprocated

rasbt replied:
sry, what would your suggested change be in this case?

jnothman replied:
All I mean is to Add a See Also section

rasbt replied:
Oh, I see! Added that one plus a reference section.

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,323 @@
+            self.subsets_[k] = {
+                'feature_subset_idx': k_idx,
+                'cv_scores': k_score,
+                'avg_score': np.nanmean(k_score)
```
Why nanmean?

rasbt replied:
This is in case some scorers might return "nan's" (I think this can happen for regressors in certain edge cases). Happy to change this to a regular np.mean() -- what do you think?

jnothman replied:
I'd think for scorers returning nans, it would be dishonest and inappropriate to just count everything else...

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,323 @@
+        Cross validation average score of the selected subset.
+
+    subsets_ : dict
+        A dictionary of selected feature subsets during the
```
Reword to make clear that it is the best feature subset of each size

rasbt replied:
done

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,323 @@
+        """
+        self._scorer = check_scoring(self.estimator, scoring=self.scoring)
+
+        if not isinstance(self.n_features_to_select, int) and\
```
Pep8 apparently tells us to avoid \-based continuations. Use () instead?

rasbt replied:
fixed

jnothman reviewed (commented):
Some comments on API. I've not looked at tests yet

rasbt commented:
Thanks a lot for the feedback, @jnothman, and no worries, it's been such a busy couple of months that I actually appreciate the delay ;). Will be addressing your comments 1-by-1 and push a new commit then!

rasbt commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,323 @@
+    Attributes
+    ----------
+    estimator_ : scikit-learn Classifier or Regressor
+        Fitted clone of `estimator`.
```
Good point, this is really not necessary. Changed that to pass the cloned estimator as function argument.

rasbt committed:
work in progress

rasbt committed:
work in progress

rasbt commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,323 @@
+    estimator_ : scikit-learn Classifier or Regressor
+        Fitted clone of `estimator`.
+
+    feature_subset_idx_ : array-like, shape = [n_predictions]
```
Sorry, can you elaborate a bit

jnothman replied:
Sorry, I meant `support_`, which is how RFE publicises its selection.

rasbt replied:
I see, but couldn't that terminology also be easily confused with "support" in context of association rule mining? I agree though that using the same terminology would be good for API consistency. How about keeping sth like `feature_subset_idx_` (which contains the indices of the subset, e.g., np.array([3, 5, 7, ...]) but also adding a `support_` mask in addition (e.g., np.array([0, 0, 0, 1, 0, 0,  1, 0, 0, 1, ...])?

rasbt commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,323 @@
+        return res
+
+    @property
+    def _estimator_type(self):
```
fixed

rasbt committed:
addressed jnothman's comments

rasbt committed:
fix unit tests

rasbt committed:
added See also and References sections to the code file

jnothman commented:
I don't see the benefit of duplicating those attributes. There's no
shortage of terminological polysemy hanging around our library, so I don't
think the argument about rule mining holds much water.

It's easy to get a list of indices from a mask with nonzero, while a list
of indices can in theory contain duplicates and be ordered, I.e. it is
over-expressive for the purpose. Similarly we could provide a sparse
diagonal components_ for the same effect... but for now I think consistency
is the most straightforward aspiration unless you have a good argument for
providing indices too.

rasbt committed:
using a support_ mask

rasbt commented:
Makes sense! Helps avoid overbloating the API and making it unnecessary complicated documentation/usage wise. Changed it to using a boolean mask as in RFE (`support_`)

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,332 @@
+        >>> sfs = sfs.fit(X, y)
+        >>> sfs.score_  # doctest: +ELLIPSIS
+        0.9733...
+        >>> sfs.feature_subset_idx_
```
needs updating

rasbt replied:
should be addressed now

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,332 @@
+        sequential feature selection algorithm.
+        Here the dictionary keys are the lengths k of these feature subsets.
+        The dictionary values are dictionaries themselves with the following
+        keys: 'feature_subset_idx' (tuple of indices of the feature subset)
```
I think this can also be called 'support' or 'support_idx'.

The formatting here won't work in HTML. Perhaps use ReST definition lists.

rasbt replied:
done

jnothman reviewed (commented):
I'll try give this another look soon

jnothman commented:
Also, can we rename the module to sequential.py?

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,332 @@
+# License: BSD 3 clause
+
+import numpy as np
+from itertools import combinations
```
we conventionally put stdlib in a first section, then external deps, then local deps

rasbt replied:
makes sense, fixed that

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,332 @@
+        of features `n_features_to_select`.
+
+    scoring : str, callable, or None (default=None)
+        A string (see model evaluation documentation) or a scorer
```
perhaps now, ```see :term:`scoring` ```

rasbt replied:
yap, that's better; fixed

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,332 @@
+    cv : int or cross-validation generator, or an iterable (default=5)
+        Determines the cross-validation splitting strategy for
+        feature selection. Possible inputs for cv are:
+        - 0 or None, don't use cross validation
```
why do we need both?

rasbt replied:
What are you exactly referring to? I copied this doc string from`cross_val_score`

jnothman replied:
Both 0 and None. Okay.

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,332 @@
+
+        """
+
+        if not (isinstance(self.n_features_to_select, int) or
```
this doesn't work for numpy scalars...

rasbt replied:
should we allow NumPy scalars if we specify `int` in the docstring? Not sure off the top of my head how that's generally handled in sklearn?!

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,332 @@
+
+        if not (isinstance(self.n_features_to_select, int) or
+                isinstance(self.n_features_to_select, tuple)):
+            raise ValueError('n_features_to_select must be a positive integer'
```
Perhaps we should just make it easy and allow the user to specify a proportion of features

rasbt replied:
Hm, I actually like the integers because they allow for being more explicit and less ambiguity (e.g., 50% of 15 features, people would have to read/check whether it would select 7 or 8 features)

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,332 @@
+        callable object / function with signature `scorer(estimator, X, y)`.
+
+    forward : bool (default=True)
+        Performs forward selection if True and backward selection, otherwise.
```
Is there a heuristic about which to use when that we should implement? Or because they're greedy, they're likely to land up finding quite different solutions (in which case, are you sure they belong in one class?)

rasbt replied:
Yeah, they can yield different solutions. In practice, I would say it depends on the dataset. If you have e.g., >10000 features and want to select 10 features, `forward=False` would be far more expensive than `forward=True`. I think it would be good to keep both options, describe the pros and cons in the docs, and let the users decide.

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,332 @@
+
+        if select_in_range:
+            max_score = float('-inf')
+            for k in self.subsets_:
```
Why can't we use range here?

rasbt replied:
instead of `for k in self.subsets_:`? what would be the advantage?

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,332 @@
+
+    def _calc_score(self, X, y, indices, estimator):
+        if self.cv:
+            scores = cross_val_score(estimator,
```
what sucks is that we can't pass in the Parallel object so it'll spin up all the forks in each iteration. Yuck. @lesteve, help!

rasbt replied:
You mean that the parallelism happens inside cross_val_score instead of parallelizing the iterations over the feature subsets? Wouldn't be hard to change and was done in the original implementation in mlxtend meanwhile: https://github.com/rasbt/mlxtend/blob/master/mlxtend/feature_selection/sequential_feature_selector.py#L407 

Let me know if you prefer this version (reg. parallelism) in scikit-learn as well

jnothman replied:
I'm not sure. I expect this could make things much faster in some setups (benchmark it? but on Windows too...). I think it would not hurt to duplicate some of the `cross_val_score` logic here., reusing `_fit_and_score`. I think the duplicated code would be quite minimal.

rasbt replied:
As far as I remember, it sped up the mlxtend implementation quite a bit. But it really depends on so many variables, thus it's hard to really benchmark it for the average use case. I think the subset-parallelism version is almost always better because you have to spawn subprocesses less often. Also, if cv < n_processors, it would allow for utilizing more resources (except there are maybe just a handful of features, and in this case, performance wouldn't matter much). I will have a look at `_fit_and_score` when I rewrite this

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,332 @@
+                                 ' n_features_to_select value.')
+
+        if isinstance(self.n_features_to_select, tuple):
+            select_in_range = True
```
Unless you really think it's necessary, I think having separate code paths for select_in_range just makes it harder to read. Can't we just set min and max to the int?

rasbt replied:
yes, I will put that on my todo list. I would say, let's revisit that once all the unit tests are in place to make sure I don't mess anything up during the rewrite :P

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,332 @@
+            for feature in remaining:
+                new_subset = tuple(subset | {feature})
+                cv_scores = self._calc_score(X, y, new_subset, estimator)
+                all_avg_scores.append(np.nanmean(cv_scores))
```
do we really want to just disregard a NaN if we get a NaN?
Should we support `error_score` as in GridSearchCV?

rasbt replied:
Yeah, using an error_score param (to either raise a warning or error upon encountering NaNs) is a good idea!

jnothman replied:
We are currently changing the default error_score in GSCV to np.NaN. We might consider using that default here already.

rasbt replied:
sure, makes sense, will do that

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,332 @@
+        remaining = orig_set - subset
+        if remaining:
+            for feature in remaining:
+                new_subset = tuple(subset | {feature})
```
Unless you correct for it elsewhere, if we are reporting the list of indices in an Attribute, I think it will have nondeterministic order due to this. Please test.

rasbt replied:
we could change it to `new_subset = tuple(sorted(subset | {feature}))`, that should fix the issue, right?

jnothman commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,332 @@
+                               X[:, indices], y)])
+        return scores
+
+    def _inclusion(self, orig_set, subset, X, y, estimator):
```
can we refactor make this and exclusion just generators of feature subsets? I.e.
```py
    def _generate_inclusion(self, orig_set, subset):
        remaining = orig_set - subset
        if remaining:
            for feature in remaining:
                yield tuple(subset | {feature})

    def _generate_exclusion(self, feature_set, fixed_feature=None):
        n = len(feature_set)
        if n > 1:
            for p in combinations(feature_set, r=n - 1):
                if fixed_feature and fixed_feature not in set(p):
                    continue
                yield p
```

rasbt replied:
will put that onto the todo list (once all unit tests are in place)

jnothman reviewed (commented):
Please add to doc/modules/classes.rst, then we need to take a look at the rendered docs.

Needs narrative docs. Does this deserve an example plot?

I'm yet to look at tests.

rasbt committed:
rebase

rasbt committed:
change to sequential.py plus docstring example fix

rasbt committed:
use double tickmarks for monospace

rasbt committed:
update whats_new

rasbt committed:
change n_features_to_select to n_features

rasbt committed:
import packages and libs in order

rasbt committed:
add see :term:`scoring`

rasbt commented on doc/whats_new.rst:
```diff
@@ -20,3 +20,5615 @@ Previous Releases
     Version 0.14 <whats_new/v0.14.rst>
     Version 0.13 <whats_new/v0.13.rst>
     Older Versions <whats_new/older_versions.rst>
```
done

rasbt commented on sklearn/feature_selection/sequential_feature_selector.py:
```diff
@@ -0,0 +1,332 @@
+        will fit a clone of the original estimator that
+        will be stored in the class attribute `self.estimators_`.
+
+    n_features_to_select : int or tuple (default=1)
```
no problem, changed it

rasbt commented:
Current todos (to keep track of what I need to do next based on the last round of feedback from @jnothman )

- [ ] Add documentation & examples
- [ ] Add more unit tests (e.g., GridSearchCV)
- [ ] potentially change parallelism (from cross_val_score to feature subsets)
- [ ] rewrite to use separate code paths for select_in_range (set max and min to ints)

jnothman commented on sklearn/feature_selection/sequential.py:
```diff
@@ -0,0 +1,334 @@
+    n_features : int or tuple (default=1)
+        An integer arguments specifies the number of features to select,
+        where n_features < the full feature set.
+        Optionally, a tuple containing a min and max value can be provided
```
Optionally -> alternatively

jnothman commented on sklearn/feature_selection/sequential.py:
```diff
@@ -0,0 +1,334 @@
+        In all other cases, ``KFold`` is used.
+
+    n_jobs : int (default=1)
+        The number of CPUs to use for cross validation.
```
See ```:term:`the Glossary <n_jobs>` ```

jnothman commented on sklearn/feature_selection/sequential.py:
```diff
@@ -0,0 +1,334 @@
+
+    subsets_ : dict
+        A dictionary containing the best selected feature subset
+        for each feature subset size selected by the
```
Selected -> evaluated

jnothman reviewed (commented):
- [ ] hiSorry

 Sorry for the incompleteness of my reviews. Why are there two test files?
Please add at least narrative docs, and perhaps an example (or modify one) comparing this to other feature selection methods in terms of runtime and optimality.

amueller commented:
no-one is working on this right now, right?

rasbt commented:
I don't think anyone is working on it right now. Somehow, I lost track of this. Is there still interest in adding it? I think we were almost done back then, only docs and tests were missing

amueller commented:
I'm definitely still interested in adding this!

rasbt commented:
Same here! Seriously lost track of it as some point. I can get back to it mid august (currently traveling)

amueller commented:
Awesome, thanks!

NicolasHug commented:
Hi @rasbt , if that's OK with you I'll start finalizing this soon in another PR. But please LMK if you'd still want to finish it! Cheers

rasbt commented:
No worries, that would be okay with me. I am really busy these days & I feel bad for not finishing this, and I'd actually appreciate it if someone else could :)
